AD400x add suport for many chips

machschmitt opened the pull request:
This PR migrates support for ADAQ4003 from the [ad_pulsar](https://github.com/analogdevicesinc/linux/blob/dev/pulsar_ardz/drivers/iio/adc/ad_pulsar.c#L736) driver to the [ad400x](https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/ad400x.c) driver.

The migration is justifiable because ADAQ4003 and AD400x devices share the same register read/write protocol, configuration register layout, and many other characteristics.

This set also aims to provide support for ADAQ400x products through HDL projects and device drivers. The request at the moment is to support ADAQ400x devices on Digilent CoraZ7 boards and have adjustable sample rate when interfacing with the ADI SPI engine IP.

To satisfy those requirements, this series is based on the [pulsar_adc_pmdz](https://github.com/analogdevicesinc/hdl/tree/master/projects/pulsar_adc_pmdz) HDL project that sources/includes the ADI SPI engine and a PWM generator IP that allows to dynamically control the sampling frequency. 

On the Linux driver side, the most important change is the addition of code for controlling the PWM device bound to the SPI engine trigger that sets the sampling frequency (implementation shamelessly stolen from [ad_pulsar](https://github.com/analogdevicesinc/linux/pull/2017/commits/b6a3bc5588da9275fee8cc1604d1122c2c05298c#diff-e73b6be5dfb0108dab447ecf4cf9483f16c48640b229c93c0a29e1e6cb68d27aR402) :)). The last patch adds devicetree file for using ADAQ4003 with CoraZ7 boards.

nunojsa commented:
@machschmitt Just making sure you're not yet looking for a review/feedback on this one? Since it's still a draft, I have been "ignoring" it...

machschmitt commented:
> @machschmitt Just making sure you're not yet looking for a review/feedback on this one? Since it's still a draft, I have been "ignoring" it...

Sure, no need to review this one. This has also not been runtime tested so I pushed it as a draft just to get the CI check it.
I'll come back to this one (and to the other PR (AD7091)) when I have a way to properly test the changes.

pdp7 commented:
@machschmitt I am tying to use the oscilloscope app with the ad4020 but [I get the error](https://github.com/adi-ses/sea-misc/pull/27#issuecomment-1747971376):
```
Failed to check if device: ad4020 has trigger. Error:No such file or directory
```
I am wondering if you have tried the osc with the ADAQ4003 and, if so, did it work?

nunojsa commented:
> Failed to check if device: ad4020 has trigger. Error:No such file or directory

That error should not matter (if really there's no triggered buffering in your device). For example, if you have HW buffering you'll see that error. That message was introduced not that long ago and I agree it is misleading and, likely, should be removed or at least refactored somehow...

machschmitt commented:
Hello @pdp7,

I often work on side branches when testing and then update pull request branches when I need feedback or think work is ready to merge. I avoid updating PR branches frequently so reviewers don't get overwhelmed with notification and I don't need to write lots of change logs.
For the ADAQ4003 project I was mainly developing on the [testing/adaq4003](https://github.com/machschmitt/linux/tree/testing/adaq4003) branch. Anyway, will update this PR as soon as I tidy code up.
I made some [tests](https://confluence.analog.com/display/~MSchmitt/ADAQ4003+tests) with pulsar hdl and linux driver from the testing branch and could get some reasonable results despite some datasheet timing requirements not being met.
One of the experiments I made was sampling a 20 kHz square wave in CH1 and a 200 kHz sine wave in CH2 at both 1MSPS and 2MSPS .
At 1MSPS we can barely see what happens when the square wave transitions from low to high.
![adaq4003_1MSPS_200kHz_sine_n_sqr](https://github.com/analogdevicesinc/linux/assets/123473008/0b031c33-3b77-47a3-9a78-12f27ce6cc88)
At 2MSPS the resulting wave form is much clear and the transition effect that happens from square wave low to high is clear.
![adaq4003_2MSPS_200kHz_sine_n_sqr](https://github.com/analogdevicesinc/linux/assets/123473008/953fe3b4-e7ea-46da-9562-21a6738ff827)

This result suggests that we are indeed reaching higher sampling frequencies.
When sampling at 1MSPS, we have one sample (point) each nano second.
Then, within 100 micro secs we can get 100000 samples points to plot the wave (although it doesn't look like we have such many points in the graph at 1MSPS).
Also, the 200kHz signal is close to the Nyquist frequency which is 500kHz when sampling at 1MSPS.
When sampling at 2MSPS, we are (presumably) sampling twice as much points, thus we see a better shaped wave.

pdp7 commented:
@machschmitt I have a MX100x to generate signal and AD4020 connected to a Zedboard.  I am expecting to see a sine wave but I am seeing a heavily quantized waveform.  I tried a variety of sine wave input frequencies and a variety of AD4020 sampling frequencies.
![Screenshot from 2023-10-06 13-25-38](https://github.com/analogdevicesinc/linux/assets/313864/aef6a614-cbd0-4db6-a0d9-9334d432cc08)
![Screenshot from 2023-10-06 13-19-31](https://github.com/analogdevicesinc/linux/assets/313864/2b4199da-47e3-41ff-90c2-db25f216c547)
![Screenshot from 2023-10-06 13-19-17](https://github.com/analogdevicesinc/linux/assets/313864/e15434be-3d00-4177-9ec2-5ce1ac69af85)
![Screenshot from 2023-10-06 13-16-50](https://github.com/analogdevicesinc/linux/assets/313864/7c9192b5-8c3e-40d6-a140-38ac36eb078b)

Do you have any of the AD40xx parts that you could test?  

It would be very helpful to understand if the heavily quantized waveform is due to a problem with the hardware setup in our office, or whether its a problem with the ad400x driver.

I opened https://github.com/machschmitt/linux/pull/1 which adds the additional AD40xx parts on top of your ADAQ4003

machschmitt commented:
@pdp7, unfortunately, I don't have any ad400x chip with me. I have ad4030 but that's completely different. The closest thing to ad400x I have is adaq4003. From [Paul's reply](https://github.com/adi-ses/sea-misc/pull/27#issuecomment-1752802724), it looks like hdl + linux driver should work with ad400x devices.

Thanks for the PR. I have cherry-picked a few changes for ad400x from [adi-ad400x-sample-freq](https://github.com/pdp7/adi-linux/tree/adi-ad400x-sample-freq). Yet, there should be more to add to properly support ad400x (I'm sure I forgot to pick the dt changes at least). There is also some cleaning to be done. Since that was a debug/test branch I have added some prints and stuff that need to be removed before merging.

Please, add whatever is needed to support ad400x devices and any additional changes you want to make if any. I'll pick up the changes, clean up any leftover test/debug code, organize the commits, and update one of the PRs for AD400x/ADAQ400x support.

Meanwhile, I'll work on other projects so we don't rework each other's changes.

pdp7 commented:
@machschmitt It turns out the MX1002 was the problem. We've connected a normal benchtop function generator to the AD4020 and the [IIO scope now shows a true sine wave](https://github.com/adi-ses/sea-misc/pull/27#issuecomment-1753171484).  I'll followup about the code changes in PR https://github.com/machschmitt/linux/pull/1

pdp7 commented:
@machschmitt do you know what the plan is for merging https://github.com/analogdevicesinc/linux/tree/staging/ad400x-add-parts ?

machschmitt commented:
> @machschmitt do you know what the plan is for merging https://github.com/analogdevicesinc/linux/tree/staging/ad400x-add-parts ?

I think it's just a matter of merging this PR.

threexc commented:
What do we need now to merge this? Still waiting on a review?

nunojsa commented:
> What do we need now to merge this? Still waiting on a review from @nunojsa ?

For the first question... to have it upstreamed (without offload support) :). For the second, I'll have a look later today

nunojsa commented on drivers/iio/adc/ad400x.c:
```diff
@@ -316,10 +316,10 @@ static int ad400x_single_conversion(struct iio_dev *indio_dev,
 
 	st->bus_locked = false;
 	spi_bus_unlock(st->spi->master);
+	iio_device_release_direct_mode(indio_dev);
```
I think this can actually be treated as a fix (an then come in the first patches) as it could lead to real deadlock

nunojsa reviewed (commented):
@threexc, review done

dlech commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -213,22 +213,20 @@ static void spi_engine_gen_sleep(struct spi_engine_program *p, bool dry,
+	t = DIV_ROUND_CLOSEST_ULL(t, 1000000000);
 	while (t) {
-		unsigned int n = min(t, 256U);
+		unsigned int n = min_t(unsigned int, t, 256U);
```
The issues with rounding here were fixed upstream. (Although in a different way.)

machschmitt commented:
Hi @nunojsa, I made all changes suggested.

Change log V1 -> V2
- Referenced adc.yaml and spi-peripherals-props.yaml in yaml dt-doc.
- Returned -ENODEV where appropriate.
- Added missing of_device_id for ad4000.
- Slipted patch adding new parts into two (one patch for chip_info, another for new parts).
- Used dev_clk_get_enabled() to simplify.
- Disabled resources in buffer enable reverse order in ad400x_buffer_postdisable.

Yet, I would say there is no need to review this again. 
Most of these commits will be squashed when upstreaming plus I have messed up with some of them after rebasing so many times.
By the way, let's work upstream.
I'll rearrange the driver and make a patchset to upstream it without spi-engine support for now.

nunojsa commented:
@machschmitt, can you rearrange this pull so it's in shape to be merged/reviewed (based on the offline discussion)?

nunojsa commented:
> It appears that the channel offset is hard-coded at -65536 and can't be set with iio_attr 

Often offset is not something you can really set (but it needs to be double checked for these devices). Now, if the different devices supported in the driver have different offsets, that needs to be addressed. Probably by setting eacho device offset in the chip_info structure.

machschmitt commented:
Hi @nunojsa, @threexc, I fixed the offset problem reported and rearranged the commits.
Hope it looks a bit better now.
It's not as good as upstream Linux patch standards but they might provide good enough support for ad400x while I get to upstream it.

Change log V2 -> V3
- Made offset writable and it now defaults to zero.
- Sign extend to return negative readings if ADC output code is twos complement.
- Squashed some fix commits I had in V2 and rearranged the remaining commits to
  put fixes early and spi-engine offload related later.

nunojsa commented on drivers/iio/adc/ad400x.c:
```diff
@@ -327,23 +344,21 @@ static void ad400x_regulator_disable(void *reg)
-	}
+	chip = (const struct ad400x_chip_info *)device_get_match_data(&spi->dev);
+	if (!chip)
+		return -EINVAL;
```
Why the removal of `spi_get_device_id()`? If this works, fine by me as eventually we should change to `spi_get_device_match_data()`

machschmitt replied:
Thought we would prefer to use `device_get_match_data()` in newer kernel versions so I updated to that while fixing that bug. Do you want me to that patch into two commits?

nunojsa commented on drivers/iio/adc/ad400x.c:
```diff
@@ -227,7 +247,7 @@ static int ad400x_read_sample(struct ad400x_state *st, uint32_t *val)
 		return ret;
 
-	*val = (input[2] << 16) | (input[1] << 8) | input[0];
+	memcpy(val, &st->data[0], 4);
```
don't we need to care about endianism?

machschmitt replied:
Indeed, maybe the best thing would be to make st->data a uint32_t variable and use `be32_to_cpu` instead of having an array of bytes for transfer buffers. I didn't want to do that change now because I would have to test buffered readings all again. I looked at drivers/iio/adc/ad4630.c to make this change, though.

machschmitt replied:
Readings I got yesterday were not obviously wrong, though not exactly the expected values.
I'll see if I can get some 15V external vref supply to see if I get better readings.

nunojsa commented on drivers/iio/adc/ad400x.c:
```diff
@@ -413,9 +413,14 @@ static int ad400x_write_raw(struct iio_dev *indio_dev,
 	case IIO_CHAN_INFO_SAMP_FREQ:
 		return ad400x_set_sampling_freq(indio_dev, val);
+	case IIO_CHAN_INFO_OFFSET:
+		st->read_offset = val;
```
Hmm this definitely feels it does not belong in here but in devicetree...

machschmitt replied:
To be honest, I'm not familiar with the use case for _offset attribute. Will look at how other drivers have that. From ABI doc (https://git.kernel.org/pub/scm/linux/kernel/git/jic23/iio.git/tree/Documentation/ABI/testing/sysfs-bus-iio?h=testing#n453) it looks like we should allow users to set that if an offset can be applied to readings. If I'm not misunderstanding things, setting a non-zero offset could  be useful of differential ADCs set up for single-ended input with IN- not grounded as example in figure 2b of SAR input types cheat sheet (https://www.analog.com/media/en/technical-documentation/product-selector-card/2pb_sarinputtypesfb.pdf).

nunojsa replied:
Yes, but not random values. The offset should indeed depend on vref and that is all about HW setup and so, devicetree :). If you can really change the offset, then I would expect your write to actually do something on the device and not just cache the value so we can read it back. Alternatively, some devices actually allow you to change the scale at runtime (even though sometimes that is also something that goes on DT) and consequently the offset might change.

nunojsa replied:
Anyways, something for you to re-think when sending upstream.

nunojsa reviewed (commented):
Some minor comments. Let me know if there's something you want to re-work before 
merging this.

I'm ok with the offset stuff but I really suspect (as it is) it's upstreamable...

mhennerich commented:
ok - let's address the remaining talking points later
